api/src/main/java/io/minio/MinioClient.java

```diff
@@ -913,32 +916,56 @@ private HttpResponse execute(Method method, String region, String bucketName, St
       request = Signer.signV4(request, region, accessKey, secretKey);
     }
 
-    if (this.traceStream != null) {
-      this.traceStream.println("---------START-HTTP---------");
-      String encodedPath = request.httpUrl().encodedPath();
-      String encodedQuery = request.httpUrl().encodedQuery();
-      if (encodedQuery != null) {
-        encodedPath += "?" + encodedQuery;
+    Response response = null;
+
+    // Loop to retry with new location of 307.
+    for (int i = 0; i < TEMP_REDIRECT_TRY_COUNT; i++) {
+      if (this.traceStream != null) {
+        this.traceStream.println("---------START-HTTP---------");
+        String encodedPath = request.httpUrl().encodedPath();
+        String encodedQuery = request.httpUrl().encodedQuery();
+        if (encodedQuery != null) {
+          encodedPath += "?" + encodedQuery;
+        }
+        this.traceStream.println(request.method() + " " + encodedPath + " HTTP/1.1");
+        String headers = request.headers().toString()
+            .replaceAll("Signature=([0-9a-f]+)", "Signature=*REDACTED*")
+            .replaceAll("Credential=([^/]+)", "Credential=*REDACTED*");
+        this.traceStream.println(headers);
+      }
+
+      response = this.httpClient.newCall(request).execute();
+      if (response == null) {
+        if (this.traceStream != null) {
+          this.traceStream.println("<NO RESPONSE>");
+          this.traceStream.println(END_HTTP);
+        }
+        throw new NoResponseException();
       }
-      this.traceStream.println(request.method() + " " + encodedPath + " HTTP/1.1");
-      String headers = request.headers().toString()
-          .replaceAll("Signature=([0-9a-f]+)", "Signature=*REDACTED*")
-          .replaceAll("Credential=([^/]+)", "Credential=*REDACTED*");
-      this.traceStream.println(headers);
-    }
 
-    Response response = this.httpClient.newCall(request).execute();
-    if (response == null) {
       if (this.traceStream != null) {
-        this.traceStream.println("<NO RESPONSE>");
-        this.traceStream.println(END_HTTP);
+        this.traceStream.println(response.protocol().toString().toUpperCase() + " " + response.code());
+        this.traceStream.println(response.headers());
+      }
+
+      if (response.code() != 307) {
+        break;
       }
-      throw new NoResponseException();
+
+      // Handle 307 (temporary redirect) specially.
+      String location = response.headers().get("Location");
+      if (location == null || location.isEmpty()) {
+        throw new InternalException("unexpected empty value in location header");
```

Comment:
IMO, this is a server issue, not internal to the SDK. Is `InternalException` justified here?